### ISSUE TITLE
[api] Parameter pull != 1 gets ignored

### ISSUE THREAD

**Issue by glogiotatidis** [docker/docker#13631]
I'm working to add `--pull` command to `docker-compose` and I found out that `dockerpy` by default sets `pull=True` (https://github.com/docker/docker-py/blob/master/docker/client.py#L309) on build. So this should be pulling the latest image always but instead it ignores pull parameter entirelly.

Digging deeper I found out that setting `pull=1` does work. So for some reason docker server does not recognize `pull=True` as `true` but only `pull=1`. Interestingly the same does not apply for `forcerm=True` which works as promised.

Here is the HTTP request of docker compose `1.3.0 RC1`

`POST /v1.18/build?pull=True&nocache=False&q=False&t=h_web&forcerm=False&rm=True HTTP/1.1`

this does not pull the image. Change `pull=True` to `pull=1` works.

For reference `docker build --pull .` sends `pull=1`
`POST /v1.18/build?cpusetcpus=&cpushares=0&dockerfile=Dockerfile&memory=0&memswap=0&pull=1&rm=1&t= HTTP/1.1`

**Comment by glogiotatidis** [comment]
```
root@docker:~# docker --version
Docker version 1.6.2, build 7c8fca2
```

```
root@docker:~# docker info
Containers: 6
Images: 14
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 26
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-52-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 1
Total Memory: 490 MiB
Name: docker
ID: XNUU:TOHF:25K7:ONTY:XBH6:TILX:4GR3:3RQH:HJH6:UBKM:3ML2:DLBJ
```

```
root@docker:~# uname -a
Linux docker 3.13.0-52-generic #86-Ubuntu SMP Mon May 4 04:32:59 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by thaJeztah** [comment]
Thanks for reporting! I think I recall some changes being made to boolean values, but I may be mistaken (that those changes were only for *presenting* the values).

Searching for PR's, I found this; https://github.com/docker/docker/pull/622/files, but that looks to be quite a while ago :smile:

I'll check if one of the maintainers knows if this has been fixed for the coming release (1.7), otherwise see if this can be fixed.

**Comment by runcom** [comment]
I looked into that early today but I'm on my mobile and cannot search in deep. The code in `postBuild` function of `api/server/server.go` was updated when we removed `engine/` and relevand code that checks booleans is here https://github.com/docker/docker/blob/master/api/server/form.go#L9

I'll have a look later today possibly

**Comment by runcom** [comment]
@glogiotatidis could you post full output of `docker version` (not `--version`)
seems like the only code path for setting `pull` to `false` is here https://github.com/docker/docker/blob/master/api/server/server.go#L1196

**Comment by runcom** [comment]
ok found

seems like we didn't maintain the old behavior but it should be good to have extra way of defining bool value in query string in the upcoming 1.7.0 release

old code was explicitly checking for pull == 1:
https://github.com/docker/docker/blob/v1.6.2/api/server/server.go#L1073

new code has a wider range of value:
https://github.com/docker/docker/blob/master/api/server/server.go#L1196

seems consistent

**Comment by runcom** [comment]
seems like it was a bug so maybe don't need a bc check at all ping @tiborvass @LK4D4

**Comment by thaJeztah** [comment]
> seems like it was a bug so maybe don't need a bc check at all 

@runcom I think you're right; if the 1.7 version *does* accept 1/True/true values (that's the expected behavior). Must have been a bug in 1.6 then and now fixed on master

**Comment by runcom** [comment]
To sum this up:

- This `pull` param was added in `1.4.0` (99.99999% sure cause 1.3 didn't have it https://github.com/docker/docker/blob/v1.3.3/api/server/server.go)
- When added `pull` was true in the server only when `if r.FormValue("pull") == "1" && version.GreaterThanOrEqualTo("1.16") {` and was like this till 1.6.2 (https://github.com/docker/docker/blob/v1.6.2/api/server/server.go#L1073)
- Now `pull` is true when `boolValue` is true (and version check) https://github.com/docker/docker/blob/master/api/server/server.go#L1197 meaning it has a wider range of value that can make pull true

**Comment by cpuguy83** [comment]
I think this is fine as is in terms of stuff needed for 1.7

**Comment by runcom** [comment]
but backward compatibily is not guaranteed anymore cause now pull can be true with values other than just "1"

**Comment by thaJeztah** [comment]
I'm +1 to use the new behavior; it should be backward compatible (1==true, 0==false). 

So, in my opinion, this issue can be closed as it is "fixed" on master. The only issue is that (I think, typing on mobile right now) the docs right now don't match the actual behavior in 1.6, but that will be solved as soon as 1.7 is released


@glogiotatidis this will mean that the feature you're working on in docker-compose will only work with Docker 1.7, unless the library you're using is patched to send `1/0` in stead of `True/False`

**Comment by glogiotatidis** [comment]
> @glogiotatidis this will mean that the feature you're working on in docker-compose will only work with Docker 1.7, unless the library you're using is patched to send 1/0 in stead of True/False

Got it, makes sense. I'll work on a patch for `docker-py` for versions <1.7. Thanks for looking into this :)

**Comment by thaJeztah** [comment]
I think we can close this. This issue is already resolved in 1.7, so not a regression, but an existing bug in 1.6,
